geolocation.py

```diff
@@ -20,27 +72,36 @@ def get_country_from_location(location_string):
     Return:
         str: a country
     """
-    country = "None"
+    # TODO: make all checks lowercase
     if location_string is None:
-        country = "None"
-    else:
-        # Loop through different typical separators of city, country, etc.
-        for separator in [",", " "]:
-            # Check different positions of the token
-            for position in [-1, 0]:
-
-                pieces = location_string.split(separator)
-                token = pieces[position].strip()
-
-                # Use returns as a way of exiting double loop
-                if token in ALL_COUNTRIES:  # pylint: disable=no-else-return
-                    return token
-                elif token in CITY_COUNTRY_DICT.keys():
-                    return CITY_COUNTRY_DICT[token]
-                elif token in CODE_COUNTRY_DICT.keys():
-                    return CODE_COUNTRY_DICT[token]
-                elif token in STATE_NAMES or token in STATE_ABBREV:
-                    return "United States"
-
-    # if no matches are found, will return "none"
-    return country
+        return "None"
+
+    # check if city,country is recognized (global and USA) as major city
+    stripped_location = location_string.replace(",", "").replace(" ", "")
+    if stripped_location in CITY_COUNTRY_STRINGS.keys():
+        return CITY_COUNTRY_STRINGS[stripped_location]
+
+    # if not international, likely to be USA: check if ends in a state
+    for state in STATE_NAMES:
+        if location_string.endswith(state):
+            return "United States"
+    for state in STATE_ABBREV:
+        if location_string.endswith(state):
+            return "United States"
+
+    # Loop through different typical separators of city, country, etc.
+    for separator in [",", " "]:
+        # Check different positions of the token
+        for position in [-1, 0]:
+            pieces = location_string.split(separator)
+            token = pieces[position].strip()                
+
+            # Use returns as a way of exiting double loop
+            if token in CITY_COUNTRY_DICT.keys() and token != 'San': # Mali has a city named San, which messes this up
```

Comment:
@kdobolyi, what do you think about just going back just dropping 'San' form the country list? I don't like the mental complexity we're adding to accomodate the city of San. I defer to you. If you were on pager duty for GitGeo, and no one will ever be on pager duty for GitGeo, what would you want?

Reply:
I'd rather not edit the list; if we stay we check all cities with at least X population, then it starts to add mental complexity to make that list. Also, by flagging San, it raises awareness that there might be other conflicts we need to think about. I don't think silently removing a city (or cities) is the way to go, because that could lead to later debugging surprises, though I admit the chances of that are remote.

Reply:
That makes sense! But you're on pager duty now. HA